demos/multicomponent/multicomponent.py.rst

```diff
@@ -0,0 +1,623 @@
+    L_h = FunctionSpace(mesh, "R", 0)               # Lagrange multiplier space
+
+    # The mixed space
+    Z_h = J_h * J_h * V_h * U_h * U_h * P_h * X_h * X_h * R_h * L_h * L_h
+    PETSc.Sys.Print(f"Mesh has {mesh.num_cells()} cells, with {Z_h.dim()} finite element DOFs")
+
+    # The trial functions
+    solution = Function(Z_h)
+    J_1, J_2, v, mu_aux_1, mu_aux_2, p, x_1, x_2, rho_inv, l_1, l_2 = split(solution)
+
+    # Lagrange multiplier trick for enforcing integral constraints
+    mu_1 = mu_aux_1 + l_1
+    mu_2 = mu_aux_2 + l_2
+
+    # The test functions
+    W_1, W_2, u, w_1, w_2, q, y_1, y_2, r, s_1, s_2 = TestFunctions(Z_h)
+
+Note that we decompose the chemical potentials as :math:`\mu_i = \mu_{i, \textrm{aux}} + l_i`
+where :math:`l_i` are Lagrange multipliers.
+This will aid in enforcing integral constraints on the solution; see
+:doc:`the steady Boussinesq demo <boussinesq.py>` for an explanation of the process.
+
+Governing PDEs: SOSM Equations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Momentum transport is modelled using the steady compressible 
+Stokes momentum equation for a Newtonian fluid,
+
+.. math::
+    :name: eq:stokes
+
+    -\nabla \cdot \big\{
+        2 \eta \epsilon(v) + (\zeta - 2\eta / d) (\nabla \cdot v) \mathbb{I}
+    \big\} + \nabla p = 0 \quad \textrm{in}\ \Omega.
+
+Recall that :math:`v` is the barycentric velocity and :math:`p` the pressure.
+Moreover :math:`\epsilon (v)` denotes the symmetric gradient of :math:`v`
+while :math:`\eta, \zeta > 0` are the shear and bulk viscosities respectively,
+:math:`d=2` is the spatial dimension and
+:math:`\mathbb{I}` the :math:`d \times d` identity matrix.
+
+We shall non-dimensionalise all of the unknowns in our discretisation.
+Hence we introduce a reference velocity :math:`v^{\textrm{ref}}` whose value
+will be specified later on when we introduce boundary conditions.
+We also choose a reference length of :math:`L^{\textrm{ref}} = 2 \cdot 10^{-3}` m.
+It is then natural to define the reference pressure as
+:math:`p^{\textrm{ref}} = \eta \cdot v^{\textrm{ref}} / L^{\textrm{ref}}`. ::
+
+    # The (dimensional) Stokes viscosities
+    eta = Constant(6e-4)            # Shear viscosity, Pa s
+    zeta = Constant(1e-7)           # Bulk viscosity, Pa s
+
+    # Non-dimensionalised Lamé parameter, dimensionless
+    lame_ND = (zeta / eta) - 1.0
+
+    # Reference quantities used for non-dimensionalisation
+    v_ref = Constant(0.0)           # Reference velocity (to be specified later), m / s
+    L_ref = Constant(2e-3)          # Reference length, m
+    p_ref = eta * v_ref / L_ref     # Reference pressure, Pa
+
+The non-dimensionalised variational formulation of :ref:`eq. 1 <eq:stokes>` is then::
+
+    # The Stokes viscous terms
+    stokes_terms = 2.0 * inner(sym(grad(v)), sym(grad(u))) * dx
+    stokes_terms += lame_ND * inner(div(v), div(u)) * dx
+
+    # The Stokes pressure term
+    stokes_terms -= inner(p, div(u)) * dx
```

Comment:
Since you mentioned Stokes, I would expect a constraint on the divergence, so a term like `-inner(div(v), q)*dx`. Maybe it'd be good to add a comment on why you only grab the momentum balance equation.